editor/editor_data.cpp

```diff
@@ -509,6 +512,138 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
+	ContextMenu cm;
+	cm.p_slot = slot;
+	cm.plugin = p_plugin;
+	p_plugin->start_idx = context_menu_plugins.size() * EditorContextMenuPlugin::MAX_ITEMS;
+	context_menu_plugins.push_back(cm);
+}
+
+void EditorData::remove_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	for (int i = context_menu_plugins.size() - 1; i > -1; i--) {
+		if (context_menu_plugins[i].p_slot == slot && context_menu_plugins[i].plugin == p_plugin) {
+			context_menu_plugins.remove_at(i);
+		}
+	}
+}
+
+int EditorData::match_context_menu_shortcut(ContextMenuSlot p_slot, const Ref<InputEvent> &p_event) {
+	for (ContextMenu &cm : context_menu_plugins) {
+		if (cm.p_slot != p_slot) {
+			continue;
+		}
+		HashMap<Ref<Shortcut>, Callable> &cms = cm.plugin->context_menu_shortcuts;
+		int shortcut_idx = 0;
+		for (KeyValue<Ref<Shortcut>, Callable> &E : cms) {
+			const Ref<Shortcut> &p_shortcut = E.key;
+			if (p_shortcut->matches_event(p_event)) {
+				return EditorData::CONTEXT_MENU_ITEM_ID_BASE + cm.plugin->start_idx + shortcut_idx;
+			}
+			shortcut_idx++;
+		}
+	}
+	return 0;
+}
+
+void EditorData::add_options_from_plugins(PopupMenu *p_popup, ContextMenuSlot p_slot, const Vector<String> &p_paths) {
+	bool add_separator = false;
+
+	for (ContextMenu &cm : context_menu_plugins) {
+		if (cm.p_slot != p_slot) {
+			continue;
+		}
+		cm.plugin->clear_context_menu_items();
+		cm.plugin->add_options(p_paths);
+		HashMap<String, EditorContextMenuPlugin::ContextMenuItem> &items = cm.plugin->context_menu_items;
+		if (items.size() > 0 && !add_separator) {
+			add_separator = true;
+			p_popup->add_separator();
+		}
+		for (KeyValue<String, EditorContextMenuPlugin::ContextMenuItem> &E : items) {
+			EditorContextMenuPlugin::ContextMenuItem &item = E.value;
+
+			if (item.icon.is_valid()) {
+				p_popup->add_icon_item(item.icon, item.item_name, item.idx);
+				const int icon_size = p_popup->get_theme_constant(SNAME("class_icon_size"), EditorStringName(Editor));
+				p_popup->set_item_icon_max_width(-1, icon_size);
+			} else {
+				p_popup->add_item(item.item_name, item.idx);
+			}
+			if (item.shortcut.is_valid()) {
+				p_popup->set_item_shortcut(-1, item.shortcut, true);
+			}
+		}
+	}
+}
+
+template <typename T>
+void EditorData::invoke_plugin_callback(ContextMenuSlot p_slot, int p_option, const T &p_arg) {
+	Variant arg = p_arg;
+	Variant *argptr = &arg;
+
+	for (int i = 0; i < context_menu_plugins.size(); i++) {
+		if (context_menu_plugins[i].p_slot != p_slot || context_menu_plugins[i].plugin.is_null()) {
+			continue;
+		}
+		Ref<EditorContextMenuPlugin> plugin = context_menu_plugins[i].plugin;
+
+		// Shortcut callback.
+		int shortcut_idx = 0;
+		int shortcut_base_idx = EditorData::CONTEXT_MENU_ITEM_ID_BASE + plugin->start_idx;
+		for (KeyValue<Ref<Shortcut>, Callable> &E : plugin->context_menu_shortcuts) {
+			if (shortcut_base_idx + shortcut_idx == p_option) {
+				const Callable &p_callable = E.value;
```

Comment:
```suggestion
				const Callable &callable = E.value;
```
Don't use `p_` for non-arguments

Reply:
Great,  done~👋